### PR TITLE
New bootstrap handshake format

### PR DESCRIPTION
**Pull request by octol**

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -28,7 +28,8 @@ pickledb = "~0.4.0"
 quick-error = "~1.2.2"
 rand = "~0.6.5"
 routing = { git = "https://github.com/maidsafe/routing.git", branch = "fleming" }
-safe-nd = "~0.4.0"
+# safe-nd = "~0.5.0"
+safe-nd = { path = "../safe-nd" }
```
Comment (octol) [inline Cargo.toml]: I'll need to update this before upgrading the draft to a PR

File: src/client_handler.rs [hunk 2]
```diff
@@ -647,33 +635,56 @@ impl ClientHandler {
...
-                {
-                    warn!("{}: Could not disconnect client: {:?}", self, err);
+    /// If a client requests the section info, we also send it.
+    fn handle_challenge(&mut self, peer_addr: SocketAddr, signature: Signature) {
+        if let Some((challenge, public_id)) = self.client_candidates.remove(&peer_addr) {
+            let public_key = match utils::own_key(&public_id) {
+                Some(pk) => pk,
+                None => {
+                    info!(
+                        "{}: Client on {} identifies as a node: {}",
```
Comment (maqi) [inline src/client_handler.rs]: this log better to be a `warn` ? also the wording can be :
`{}: Client on {} identifies as a node: {}, hence disconnect from it.`

File: src/client_handler.rs [hunk 3]
```diff
@@ -1106,6 +1117,77 @@ impl ClientHandler {
...
+                    .collect::<Vec<_>>(),
+                Err(e) => {
+                    info!("Could not handle bootstrap request: {:?}", e);
+                    return;
+                }
+            };
+
+            if closest_known_elders.is_empty() {
+                warn!(
+                    "{}: No closest known elders in any section we know of",
```
Comment (maqi) [inline src/client_handler.rs]: this sounds more like meaning this vault knows no elders ?
I'd prefer a log_or_panic here as it's very suspicious.

File: src/client_handler.rs [hunk 4]
```diff
@@ -1106,6 +1117,77 @@ impl ClientHandler {
...
+                    self
+                );
+            } else {
+                self.send(peer_addr, &HandshakeResponse::Join(closest_known_elders));
```
Comment (maqi) [inline src/client_handler.rs]: just to confirm, so this means we are sending the `HandshakeResponse::Join(closest_known_elders)` back to the client,
and he shall try to connect to them, right?

File: src/mock_routing/mod.rs [hunk 5]
```diff
@@ -94,6 +99,26 @@ impl Node {
         Ok(())
     }
 
+    /// Find out if the given XorName matches our prefix.
+    pub fn matches_our_prefix(&self, _name: &routing::XorName) -> Result<bool, RoutingError> {
```
Comment (maqi) [inline src/mock_routing/mod.rs]: better import `XorName` in the above use block ?
and same to the other places

File: tests/common/mod.rs [hunk 6]
```diff
@@ -226,28 +236,56 @@ pub trait TestClientTrait {
...
+
+        let handshake_response: HandshakeResponse = unwrap!(bincode::deserialize(&bytes));
+        let _payload = match handshake_response {
+            HandshakeResponse::Join(payload) => payload,
+            _ => panic!("Unexpected HandshakeResponse"),
+        };
+
+        // TODO: handle payload
+        // Currently we are connected to a single vault. At this stage we should disconnect from it
+        // and connect to the new elders we are given as payload.
```
Comment (maqi) [inline tests/common/mod.rs]: Shall we simulate the process even with just one vault?
I am asking this as the `self.send(&request)` actually only sends to the connected vaults.
which seems the test is not covering the whole procedure?

File: src/client_handler.rs (hunk 3 above)
Comment (octol) [reply src/client_handler.rs]: Don't think we use `log_or_panic` in this crate though

File: src/client_handler.rs (hunk 4 above)
Comment (octol) [reply src/client_handler.rs]: That's correct

File: tests/common/mod.rs (hunk 6 above)
Comment (octol) [reply tests/common/mod.rs]: This is phase 2B though for multiple sections?

File: src/client_handler.rs (hunk 3 above)
Comment (maqi) [reply src/client_handler.rs]: ok, then

File: tests/common/mod.rs (hunk 6 above)
Comment (maqi) [reply tests/common/mod.rs]: I mean, this flow indicates that if within the same section, then probably don't need to carry out the disconnection and connect to the new.
So, maybe worth to mentionin the comment that:
`Disconnect if we are not among the elders` ?

Comment (octol) [reply tests/common/mod.rs]: Good point, I'll update the comment

**Pull request merged**